### ISSUE TITLE
article.parent should not be a foreign key 

### ISSUE THREAD

**Issue by vanatteveldt** [amcat/amcat#460]
Currently, articles have a parent property that points to another article object. This is used to indicate e.g. replies to post on social media.

In the context of single AmCAT servers, having articles reference each other makes sense (although it also causes problems with scraping / uploading articles as the client doesn't know the ids or parent articles yet, for which a special purpose "children" property (list of child articles) was added to the upload API)

With multiple AmCAT servers, however, this is very problematic, especially if you wish to copy a child object without also copying the parent. Article IDs don't carry between versions, so some other identifier will be needed (e.g. uuid), but if that can't be dereferenced (because the parent is unknown) it is impossible to store the article without modifying it by removing the parent attribute. 

Proposed solution: make article.parent a "weak reference" / "symbolic link". This could be either uuid or url, buy my preference would be url, since (1) it is more meaningful, (2) the client can provide the parent and child urls as needed, and (3) I don't think accidental duplicate urls are a problem since we don't really do anything advanced with the parent field, so if the tree doesn't look nice because of dupes it's a user problem. 

This will allow for the children upload code to be removed (and maybe upload can be folded back into POST article?), as well as the new code for ordering by parent and converting parent to and from uuid. 
As far as I can see, for the rest only the article details tree code will need to be adapted.

@CJStuart @martijnbastiaan comments?

**Comment by martijnbastiaan** [comment]
I do have problems with choosing the URL as a parent field, as I've mentioned in our private chats. First of all, I do agree that it is more meaningful (from a human perspective) to have a URL as a parent field. But..

> 2) the client can provide the parent and child urls as needed

That's true for UUID as well (just follow the url of the child, and figure out the parent from the source medium). If you're thinking of the client polling the original database (amcat.nl, for example) to get the missing articles, that's certainly doable with UUIDs as well - and I suppose even easier.

>  (3) I don't think accidental duplicate urls are a problem since we don't really do anything advanced with the parent field, so if the tree doesn't look nice because of dupes it's a user problem. 

Then just don't do it all. We should either support a feature, or not support it. *Unreliable* features are way worse than not having it all. For example, 5938 of the first 863552 articles (`select url, count(url) from articles where article_id < 1000000 group by url having count(url)>1;`) have a duplicate URL. Obviously, that's not much but it's still unreliable.

> it's a user problem

Except that these issues are probably introduced by scraping, which is hardly a user thing.

Some additional points:

1. Introducing the URL as a parent forces software to guess by design. URLs are supposed to be unique (hence the name), but in practice aren't. Duplicate articles, differently scraped articles, edits, etc. are all real things. That's, I suppose, one of the reasons we don't force a unique constraint on urls right now. So this is a combination of:

 * In the face of ambiguity, refuse the temptation to guess.
 * Errors should never pass silently.

2. I cannot imagine a text field (which is probably always stored on some sort of heap, like you said before) will ever be fast to process if we have to.

3. User problems (article tree not rendered correctly) will be developers problems at some point. 

To summarize: I think we'll introduce ~~buggy~~ unreliable software from a user's perspective

**Comment by martijnbastiaan** [comment]
Whatever we decide to do, it might also be a good idea to store a 'root' column pointing to the root of the article tree. This would make getting the tree an article is in O(1), and would make it potentially much easier to import missing articles.

I.e., AmCAT could display a message 'this article is missing parents, would you like to import them?' and then simply query (various?) servers with something like `/api/v4/articletree/$root_{uuid,url}`.

**Comment by vanatteveldt** [comment]
[I think the "include root article and allow finding missing parents" should be a separate issue (and I'm not sure I think the costs are worth the benefit)]

Re client-provided UUIDs: this is not possible for a scraper without giving up deduplication - if the scraper provides a UUID, the article can't be deduplicated against existing articles anymore

Re unreliable feature: I wouldn't mind not showing the tree at all, but only displaying parent (with links to any articles with that url in the current set) and children (with links). This doesn't really make it unreliable, I think.,

Re user problem: If we scrape, we are the user :). So then it's our problem, but from a "local VU amcat" perspective

Re your other points 1 and 3: I think they are solved by not asserting that articles need to be a tree, and just displaying local context. 

Re 2: That might be a valid point, and UUIDs are probably faster than URLs in SQL. However, I think we should use elastic anyway if processing is problematic (i.e. outside the scope of single articles), and I don't think we really have plans to do anything with parents at the moment

**Comment by vanatteveldt** [comment]
This has been fixed in 3.5, parents are now parent_hash
